### ISSUE TITLE
DCGAN for Fashion MNIST

### ISSUE THREAD

**Issue by OppiHmiK** [OppiHmiK/Artificial-Intelligence#2]
import matplotlib.pyplot as plt
import numpy as np

from keras.datasets import fashion_mnist as f_mnist
from keras import Sequential
from keras.layers import Dense, Activation, Flatten, Reshape
from keras.layers import Conv2D, Conv2DTranspose, UpSampling2D
from keras.layers import LeakyReLU, Dropout
from keras.layers import BatchNormalization
from keras.optimizers import Adam, RMSprop
from keras.utils import to_categorical

(i_train, t_train), (i_test, t_test) = f_mnist.load_data()

class DCGAN(object):

    def __init__(self, img_r = 28, img_c = 28, channel = 1):

        self.img_r = img_r
        self.img_c = img_c # NOTE : 28*28 사이즈의 이미지
        self.channel = channel

        self.Discriminator = None # NOTE : 판별자
        self.Generator = None # NOTE : 생성자
        self.A_Model = None # NOTE :적대적 모델
        self.D_Model = None # NOTE :판별 모델

    def discriminator(self):

        if self.Discriminator: # NOTE : Discriminator에 값이 들어가 있으면 Discriminator 반환
                               # NOTE : 처음엔 None값이 저장되어 있으므로, if 문에 종속된 코드는 돌아가지 않음.
            return self.Discriminator

        self.Discriminator = Sequential()  # NOTE : 아닐 경우에는 모델을 순차적으로 만듦.
        depth = 64 # NOTE : input data (28 x 28 x 1, depth = 1), output data (14 x 14 x 1, depth = 64)
        dropout = 0.4
        slope = 0.2  # NOTE : 학습률

        i_shape = (self.img_r, self.img_c, self.channel)
        self.Discriminator.add(Conv2D(depth*1, 5, strides = 2, input_shape = i_shape, padding = 'same'))
        self.Discriminator.add(LeakyReLU(alpha = slope))
        self.Discriminator.add(Dropout(dropout))

        self.Discriminator.add(Conv2D(depth * 2, 5, strides=2, padding='same'))
        self.Discriminator.add(LeakyReLU(alpha=slope))
        self.Discriminator.add(Dropout(dropout))

        self.Discriminator.add(Conv2D(depth * 4, 5, strides=2, padding='same'))
        self.Discriminator.add(LeakyReLU(alpha=slope))
        self.Discriminator.add(Dropout(dropout))

        self.Discriminator.add(Conv2D(depth * 8, 5, strides=2, padding='same'))
        self.Discriminator.add(LeakyReLU(alpha=slope))
        self.Discriminator.add(Dropout(dropout))

        self.Discriminator.add(Flatten())
        self.Discriminator.add(Dense(1))
        self.Discriminator.add(Activation('sigmoid'))
        self.Discriminator.summary()
        return self.Discriminator

    def generator(self):

        if self.Generator:
            return self.Generator

        self.Generator = Sequential()
        dropout = 0.4
        momentum = 0.9
        depth = 256
        dim = 7

        self.Generator.add(Dense(dim*dim*depth, input_dim = 100))
        self.Generator.add(BatchNormalization(momentum = momentum))
        self.Generator.add(Activation('relu'))
        self.Generator.add(Reshape((dim, dim, depth)))
        self.Generator.add(Dropout(dropout))

        self.Generator.add(UpSampling2D())
        self.Generator.add(Conv2DTranspose(int(depth/2), 5, padding = 'same'))
        self.Generator.add(BatchNormalization(momentum=momentum))
        self.Generator.add(Activation('relu'))

        self.Generator.add(UpSampling2D())
        self.Generator.add(Conv2DTranspose(int(depth / 4), 5, padding='same'))
        self.Generator.add(BatchNormalization(momentum=momentum))
        self.Generator.add(Activation('relu'))

        self.Generator.add(UpSampling2D())
        self.Generator.add(Conv2DTranspose(int(depth / 8), 5, padding='same'))
        self.Generator.add(BatchNormalization(momentum=momentum))
        self.Generator.add(Activation('relu'))

        self.Generator.add(Conv2DTranspose(1, 5, padding='same'))
        self.Generator.add(Activation('sigmoid'))
        self.Generator.summary()
        return self.Generator

    def d_model(self):

        if self.D_Model:
            return self.D_Model

        learning_rate = 0.0002
        decay = 6e-8
        opti = RMSprop(lr = learning_rate, decay = decay)
        self.D_Model = Sequential()
        self.D_Model.add(self.discriminator())
        self.D_Model.compile(loss = 'binary_crossentropy', optimizer = opti, metrics = ['accuracy'])
        return self.D_Model

    def a_model(self):

        if self.A_Model:
            return self.A_Model

        learning_rate = 1e-3
        decay = 3e-8
        opti = RMSprop(lr = learning_rate, decay = decay)

        self.A_Model = Sequential()
        self.A_Model.add(self.generator())
        self.A_Model.add(self.discriminator())
        self.A_Model.compile(loss = 'binary_crossentropy', optimizer = opti, metrics = ['accuracy'])
        return self.A_Model

class f_MNIST_DCGAN(object):

    def __init__(self):

        self.img_r = 28
        self.img_c = 28
        self.channel = 1
        self.i_train = i_train
        self.i_train = self.i_train.reshape(-1, self.img_r, self.img_c, 1).astype(np.float32)

        self.DCGAN = DCGAN()
        self.discriminator = self.DCGAN.d_model()
        self.adversarial = self.DCGAN.a_model()
        self.generator = self.DCGAN.generator()

    def self_img(self, s2file = False, fake = True, sample = 16, noise = None, epoch = 0):
        filename = 'f_mnist.png'
        if fake:
            if noise in None:
                noise = np.random.uniform(-1.0, 1.0, size = [sample, 100])

            else:
                filename = "f_mnist_%d.png"%epoch

            img = self.generator.predict(noise)

        else:
            rep = np.random.randint(0, self.i_train.shape[0], sample)
            img = self.i_train[rep, :, :, :]

        plt.figure(figsize = (10, 10))
        for rep in range(img.shape[0]):
            plt.subplot(4, 4, rep + 1)
            img2 = img[rep, :, :, :]
            img2 = np.reshape(img2, [self.img_r, self.img_c])
            plt.imshow(img2, camp = 'gray')
            plt.axis('off')
        plt.tight_layout()

        if s2file:
            plt.savefig(filename)
            plt.close('all')

        else:
            plt.show()

    def train(self, epoch = 2000, b_size = 256, s_interval = 0):

        n_Input = None
        if s_interval > 0:
            n_Input = np.random.uniform(-1.0, 1.0, size = [16, 100])

        for rep in range(epoch):
            img_train = self.i_train[np.random.randint(
                0, self.i_train.shape[0], size = b_size), : ,:, :]

            noise = np.random.uniform(-1.0, 1.0, size = [b_size, 100])
            img_fake = self.generator.predict(noise)
            i_data = np.concatenate((img_train,img_fake))
            o_data = np.ones([2*b_size, 1])
            o_data[b_size:, :] = 0
            d_cost = self.discriminator.train_on_batch(i_data, o_data)

            o_data = np.ones([b_size, 1])
            noise = np.random.uniform(-1.0, 1.0, size = [b_size, 100])
            a_cost = self.adversarial.train_on_batch(noise, o_data)

            log_mes = "%d : [판별자 오차 : %f, 정확도 : %f]"%(rep, d_cost[0],d_cost[1])
            log_mes = "%s : [적대적 오차 : %f, 정확도 : %f]"%(log_mes, a_cost[0], a_cost[1])
            print(log_mes)

            if s_interval > 0:
                if (rep + 1) % s_interval ==0:
                    self.self_img(s2file = True, fake = True, sample = n_Input.shape[0], noise = n_input, epoch = (epoch + 1))

if __name__ == '__main__':
    fa_mnist_DCGAN = f_MNIST_DCGAN()
    fa_mnist_DCGAN.train(epoch = 10000, b_size = 256, s_interval= 500)
    fa_mnist_DCGAN.self_img(fake=True)
    fa_mnist_DCGAN.self_img(fake = False, s2file = True)

**Comment by OnedgeLee** [comment]
선별적으로 gradient를 apply하는 부분에 문제가 있지 않나요?
generator loss에 대한 gradient는 generator에만 적용되어야 하고,
discriminator loss에 대한 gradient는 discriminator에만 적용되어야 하는데,
generator loss의 gradient를 apply할 때, discriminator까지 함께 학습되는 것으로 보입니다.
